sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java

```diff
@@ -341,63 +327,115 @@ public UnsafeArrayData copy() {
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.BOOLEAN_ARRAY_OFFSET, size);
+    return values;
+  }
 
-    final int offsetRegionSize = 4 * arr.length;
-    final int valueRegionSize = 4 * arr.length;
-    final int totalSize = 4 + offsetRegionSize + valueRegionSize;
-    final byte[] data = new byte[totalSize];
+  @Override
+  public byte[] toByteArray() {
+    int size = numElements();
+    byte[] values = new byte[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.BYTE_ARRAY_OFFSET, size);
+    return values;
+  }
 
-    Platform.putInt(data, Platform.BYTE_ARRAY_OFFSET, arr.length);
+  @Override
+  public short[] toShortArray() {
+    int size = numElements();
+    short[] values = new short[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.SHORT_ARRAY_OFFSET, size * 2);
+    return values;
+  }
 
-    int offsetPosition = Platform.BYTE_ARRAY_OFFSET + 4;
-    int valueOffset = 4 + offsetRegionSize;
-    for (int i = 0; i < arr.length; i++) {
-      Platform.putInt(data, offsetPosition, valueOffset);
-      offsetPosition += 4;
-      valueOffset += 4;
-    }
+  @Override
+  public int[] toIntArray() {
+    int size = numElements();
+    int[] values = new int[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.INT_ARRAY_OFFSET, size * 4);
+    return values;
+  }
 
-    Platform.copyMemory(arr, Platform.INT_ARRAY_OFFSET, data,
-      Platform.BYTE_ARRAY_OFFSET + 4 + offsetRegionSize, valueRegionSize);
+  @Override
+  public long[] toLongArray() {
+    int size = numElements();
+    long[] values = new long[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.LONG_ARRAY_OFFSET, size * 8);
+    return values;
+  }
 
-    UnsafeArrayData result = new UnsafeArrayData();
-    result.pointTo(data, Platform.BYTE_ARRAY_OFFSET, totalSize);
-    return result;
+  @Override
+  public float[] toFloatArray() {
+    int size = numElements();
+    float[] values = new float[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.FLOAT_ARRAY_OFFSET, size * 4);
+    return values;
   }
 
-  public static UnsafeArrayData fromPrimitiveArray(double[] arr) {
-    if (arr.length > (Integer.MAX_VALUE - 4) / 12) {
+  @Override
+  public double[] toDoubleArray() {
+    int size = numElements();
+    double[] values = new double[size];
+    Platform.copyMemory(
+      baseObject, baseOffset + headerInBytes, values, Platform.DOUBLE_ARRAY_OFFSET, size * 8);
+    return values;
+  }
+
+  private static UnsafeArrayData fromPrimitiveArray(
+       Object arr, int offset, int length, int elementSize) {
+    final int headerInBytes = calculateHeaderPortionInBytes(length);
+    final int valueRegionInBytes = elementSize * length;
+    final int totalSizeInLongs = (headerInBytes + valueRegionInBytes + 7) / 8;
```

Comment:
Sorry I was wrong :(  We need to declare them all as long, in order to do overflow check.